lib/Optimizer/Transforms/Mapping.cpp

```diff
@@ -0,0 +1,624 @@
+ ******************************************************************************/
+
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "cudaq/Support/Device.h"
+#include "cudaq/Support/Placement.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/ScopedPrinter.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Transforms/TopologicalSortUtils.h"
+
+#define DEBUG_TYPE "quantum-mapper"
+
+using namespace cudaq;
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// Generated logic
+//===----------------------------------------------------------------------===//
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_MAPPINGPASS
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Placement
+//===----------------------------------------------------------------------===//
+
+void identityPlacement(Placement &placement) {
+  for (unsigned i = 0, end = placement.getNumVirtualQ(); i < end; ++i)
+    placement.map(Placement::VirtualQ(i), Placement::DeviceQ(i));
+}
+
+//===----------------------------------------------------------------------===//
+// Routing
+//===----------------------------------------------------------------------===//
+
+/// This class encapsulates an quake operation that uses wires with information
+/// about the virtual qubits these wires correspond.
+struct VirtualOp {
+  mlir::Operation *op;
+  SmallVector<Placement::VirtualQ, 2> qubits;
+
+  VirtualOp(mlir::Operation *op, ArrayRef<Placement::VirtualQ> qubits)
+      : op(op), qubits(qubits) {}
+};
+
+class SabreRouter {
+  using WireMap = DenseMap<Value, Placement::VirtualQ>;
+  using Swap = std::pair<Placement::DeviceQ, Placement::DeviceQ>;
+
+public:
+  SabreRouter(const Device &device, WireMap &wireMap, Placement &placement)
+      : device(device), wireToVirtualQ(wireMap), placement(placement),
+        phyDecay(device.getNumQubits(), 1.0), phyToWire(device.getNumQubits()),
+        allowMeasurementMapping(false) {}
+
+  void route(Block &block, ArrayRef<quake::NullWireOp> sources);
+
+private:
+  void visitUsers(ResultRange::user_range users,
+                  SmallVectorImpl<VirtualOp> &layer,
+                  SmallVectorImpl<Operation *> *incremented = nullptr);
+
+  LogicalResult mapOperation(VirtualOp &virtOp);
+
+  LogicalResult mapFrontLayer();
+
+  void selectExtendedLayer();
+
+  double computeLayerCost(ArrayRef<VirtualOp> layer);
+
+  Swap chooseSwap();
+
+private:
+  const Device &device;
+  WireMap &wireToVirtualQ;
+  Placement &placement;
+
+  // Parameters
+  unsigned extendedLayerSize = 20;
+  float extendedLayerWeight = 0.5;
+  float decayDelta = 0.5;
+  unsigned roundsDecayReset = 5;
```

Comment:
Exposed the 4 algorithm parameters as pass options in cd4eda08.

Reply:
I was thinking we should put `const` or `constexpr` in front of them as a qualifier, since they don't seem to be configurable.

Reply:
I made them configurable at pass creation via Pass Options, which was one of the items on the checklist in the parent issue anyway. I also made them `const` in the `SabreRouter` class since they won't change once the pass is created.